text/0055-signature-based-finality.md

```diff
@@ -0,0 +1,115 @@
+
+A node receiving the request sent in point 6 would respond with the following:
+- To the request for a single fully signed block, send the block along with signatures proving it is fully signed.
+- To the request for a range of fully signed blocks, send the fully signed blocks one by one, starting from the oldest one.
+
+Upon receiving a block with signatures as a response:
+    - If it's not fully signed, discard and re-request it from someone else.
+    - If it's the immediate descendant of the highest known block:
+        - Request execution by the contract runtime.
+        - Compare the execution results to the block - if they don't match, discard the block and re-request it from someone else.
+        - If everything fits, store the block as the new highest block.
+    - If it's not the immediate descendant of the highest known block, request the range of blocks from the immediate descendant of the highest block to this one.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The `ContractRuntimeAnnouncement::LinearChainBlock` has to be handled both by the `LinearChain` and consensus.
+
+`LinearChain`, instead of storing it immediately, would first cache it and wait for finality signatures. Only after it is fully signed and the immediate ancestor of the highest finalized block, will it be appended to the linear chain and announced in a `BlockAdded` announcement.
+
+Consensus would react to the `LinearChainBlock` announcement the way it currently handles a `BlockAdded` announcement - it would create and announce a finality signature.
+
+The `ConsensusAnnouncement::CreatedFinalitySignature` would be handled by the linear chain as before, except now the `LinearChain` component would check whether the signed block is now fully signed. If it is, it would do the checks described in the sub-points of point 6 in the [Guide level explanation](#guide-level-explanation).
+
+Two new network message variants would be added to `protocol::Message`: `RequestBlockRange` and `FullySignedBlock`.
+
+Whenever the `LinearChain` component collects enough finality signatures for a block hash, but it doesn't know the block, it needs to request that block. It can be done with the already existing `Message::GetRequest`.
+
+`RequestBlockRange(a, b)` handles the use case when a new fully signed block is not the immediate descendant of the highest known fully signed block. `a` and `b` denote the minimum and maximum requested block heights.
+
+`FullySignedBlock(block, signatures)` would be used as a response to a `RequestBlockRange`. Sometimes, multiple such responses would be required.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+This complicates the process of finalization slightly, and will probably increase time to finality (the time elapsed between when the block was proposed and when it was fully finalized, ie. fully signed). However, the significant memory savings enabled by this are worth the tradeoff.
```

Comment:
I am not sure it would increase the _time-to-finality_ since, from the perspective of the single node the block is finalized as soon as that node detects the finality. When will we produce an event to the event stream about block's finalization now?

Reply:
I see now, in the _Reference-level implementation_:

> Only after it is fully signed and the immediate ancestor of the highest finalized block, will it be appended to the linear chain and announced in a `BlockAdded` announcement.